### ISSUE TITLE
Open discussion for status and future enhancements

### ISSUE THREAD

**Issue by belgattitude** [belgattitude/php-java-bridge#6]
Here's a special issue to share ideas about the future of the project...

__To the current point (6.2.10 - Jan 17)__:
> the PHPJavaBridge showed **stable and robust**... Can successfully be compiled with **recent JDK** > (7 and 8 tested) and deployed on a **modern Tomcat** (7 and 8 tested). It also reveals possible to
> convert it on a **servlet 3.0 spec** container and **embrace modern ecosystem** (see example of the 
> [pjb-starter-gradle](https://github.com/belgattitude/pjb-starter-gradle/).
>
> The client `Java.inc` was completely refactored to support **PHP7 and HHVM** and embrace more
> more modern practices that showed up in the **PHP ecosystem (like composer, psr4 autoloading...)**
> The client project lives now in a separate repository named [soluble-japha](https://github.com/belgattitude/soluble-japha). 
>Don't use the `Java.inc`way anymore or use the compatibility layer [soluble-japha-pjb62-compat](https://github.com/belgattitude/soluble-japha-pjb62-compat) if you have to migrate code.
>
> Also note the conversion from sourceforge CVS to github which offers a better infrastructure and facilitate contributions.

So what would be the next steps ?

To my point of view, I would think about infrastucture work before implementing features: 

**1. Some refactorings** :
- [ ] Drop the Java.inc code (still bundled when building)
- [x] Convert or drop PHP scripts that were used to build the server (see ./server/*.php)
- [x] Remove the building of the `JavaBridge.war` (>40Mb) template (will help to remove completely `./unsupported` directory. This was was created as a sort of test-everything (Birt reports....) but requires tons of outdated deps. We can keep the minimal `JavaBridgeTemplate.war`.
- [x] Drop the inclusion of `script-api.jar`, its was made available for pre-Java 6.
- [x] Drop of all `./debian` directory (Too old)
- [x] Drop of all `./security` (example of SELINUX policies for Fedora 4)
- [x] Publish on Maven, make an artifact (**Anyone with experience ?**)

**2. Unit testing**:

Currently I didn't spend time to really how to run them... 

- [x] **PHPtests**: I suggest to **convert and drop PHP remaining (non php-)unit tests** found in `./test.php5` and put them in the soluble-japha client project. See some example [here](https://github.com/belgattitude/soluble-japha/tree/master/test/src/SolubleTest/Japha/Bridge) and a specific file [here](https://github.com/belgattitude/soluble-japha/blob/master/test/src/SolubleTest/Japha/Bridge/AdapterUsageCommonTest.php). I think **it's already made 90%**, still some cases to review before dropping. Reviewers welcome !

- [x] **JavaTests** Help and suggestions welcomed !! **Anybody** with time to figure out, explain and propose a solution?

**3. Quality**

- [x] Add a build target on travis (will possibly require to first drop php requirement in the build)

I think this would be the first steps before thinking about additions (had some ideas about replacing XML by [protocol-buffers](https://developers.google.com/protocol-buffers/)... fun work but let's see)

Share your ideas... or create referenced issues.

**Comment by domdorn** [comment]
the "Publish on Maven" should be doable with https://docs.gradle.org/current/userguide/publishing_maven.html 
As you're already having Gradle in the project, that should be straight forward.

**Comment by cplerch** [comment]
I would give high priority to these tasks for the Java-part of the project:

   - Split the project strictly into a Java part and a PHP part (in fact 2
   related projects); remove all PHP scripts used to build the Java-part - PHP
   shouldn't be required to buid the Java project; drop bundled Java.inc
   - Refactor existing and setup new Java unit tests in order to establish
   full functional coverage
   - Mavenize the project directory layout; this is also necessary for
   using the Gradle Java plugin for all artifacts (and not only for the webapp)
   - Modularize the Java project. Good candidates being (but needs to be
   investigated more):
      - PhpJavaBridge.jar: PHP -(call)-> Java methods
      - JavaPhpScripting.jar: Java -(exec)-> PHP functions
      - PhpJavaServlet.jar: servlet based on PhpJavaBridge.jar (perhaps
      with FastCGI removed?)
      - PhpJavaBridge.war: modern servlet api 3.0 webapp
      - PhpJavaBridgeServer.jar: perhaps a standalone version of the .war
      with embedded application server
   - Get rid of Ant: migrate all build tasks to native Gradle (or Maven:
   why not?)
   - Modernize and streamline the code; it looks convoluted and wired to me.
   - And then perhaps: optimize performance, add functionality, etc.


2017-01-18 19:21 GMT+01:00 Dominik Dorn <notifications@github.com>:

**Comment by belgattitude** [comment]
Great I'll take a look somewhere next week, most of the things seems ok for me.

If it's only about me I would completely drop the "javaPhpScripting.jar: Java -(exec)-> PHP functions" support. 

If we agree on some tasks, can you precise if you would work on it ?

**Comment by cplerch** [comment]
1. Fine.

2. Already though about this, but the looking at the code I quickly
disrearded this option. Both parts (php-->Java, Java-->php) are intertwined
and nested like Siamese twins. Of course we could try to separate them and
let one of them die...

3. Yes, yes, yes. I would like to start with getting the unit tests for
php-->Java running in the current version of the code base. Agreed?

Best,
Chris

2017-01-19 12:49 GMT+01:00 Sébastien Vanvelthem <notifications@github.com>:

**Comment by belgattitude** [comment]
Hi Chris,

> 3. I would like to start with getting the unit tests for php->Java running in the current version of the code base. Agreed?

I've created an issue (will update the description later on) and once you've confirmed as collaborator I'll assign to you... see issue #7.

Unfortunately, I haven't had the time to work on a manual for contributors yet, but you can take as an example the [contributing template for soluble-japha](https://github.com/belgattitude/soluble-japha/blob/master/CONTRIBUTING.md#recommended-workflow-for-contributions).

In short: I suggest to make your own fork (from the github page), create a branch (like - 'hotfix/7') on which you work and publish a pull-request on my repo. So I'll check on my side and merge accordingly. 

Keep in mind, that I really plan to drop the Java.inc client at some point...  So don't focus on php tests, they will be made on the soluble-japha side (currently coverage at approx 60%, but a lot of obsolete code). Of course you'll probably depends on Java.inc in some parts of the java unit tests... Could be nice to keep track of them for later conversion to soluble-japha.

Good work ;) 

*(I think also about moving to a github organisation like phpjavabridge once I have more time, but needs to take time to understand advantages and drawbacks, limitations withouts paying...)*

**Comment by belgattitude** [comment]
Sorry fixed the link for the soluble-japha example [contributor workflow](https://github.com/belgattitude/soluble-japha/blob/master/CONTRIBUTING.md#recommended-workflow-for-contributions)

**Comment by belgattitude** [comment]
Hi @cplerch and @domdorn 

A huge thank to you, looks publishing is on the way. Just released the 6.2.11-rc-1, not yet perfect but you can test it:

With maven
```xml
<dependency>
    <groupId>io.soluble.pjb</groupId>
    <artifactId>php-java-bridge</artifactId>
    <version>VERSION</version>
</dependency>
```

With gradle

```
compile 'io.soluble.pjb:php-java-bridge:6.2.11-rc-1'
```

I'll test it more before release but it' a start
